### PR TITLE
[Tests] Add unit tests for srt/multimodal/

### PR DESCRIPTION
**Pull request by jinanthony**
## Motivation, Modifications

addresses srt/multimodal/ unit tests from this issue: https://github.com/sgl-project/sglang/issues/20865

### Coverage summary (intent)

| Module | Coverage goal | Notes |
|--------|----------------|--------|
| `evs/evs_core.py` | **~100%** of executable logic | Retention math, mask, `tokens_per_frame`, `replace_offsets_with_tokens_per_frame` are all exercised. |
| `customized_mm_processor_utils.py` | **~100%** | Small registry module; register / error / overwrite paths tested. |
| `internvl_utils.py` | **~100%** when tests run | Test classes are **skipped** if `torchvision` / import fails → that file shows **0%** in that run. |
| `mm_utils.py` | **High, not full** | Preprocessing, grid, pad, DP **assignment**, `process_anyres_image`, `process_images` covered. **`run_dp_sharded_vision_model`** and **`run_dp_sharded_mrope_vision_model`** need tensor-parallel / distributed — not covered by these CPU unit tests unless separately mocked. |
| `processors/base_processor.py` | **Partial** | Pure/static paths (tokens, offsets, validation, `_process_loaded_mm_data`) via loaded source + stubs. Large runtime paths (tokenizer, executors, `load_mm_data`, CUDA IPC, NPU) are **out of scope** for this suite. |

---

### `test_mm_utils.py` → `sglang.srt.multimodal.mm_utils`

| Area | What it checks | Coverage |
|------|----------------|----------|
| **`TestEnsureNumpy`** | Tensor → `ndarray`; numpy/other pass-through. | `ensure_numpy`. |
| **`TestHasValidData`** | `None` / `[]` invalid; nested truthy valid. | `has_valid_data` + nested flatten path. |
| **`TestSelectBestResolution`** | Best effective fit; single candidate; tie-break on wasted pixels. | Full `select_best_resolution` loop + tie branch. |
| **`TestGetAnyresImageGridShape`** | List pinpoints; `"(1x1)(2x2)"` expansion; `literal_eval` string list; invalid `patch_size` asserts. | String-`x` branch, list branch, `ast.literal_eval` branch, assert path. |
| **`TestDivideToPatches`** | Even grid, one patch, uneven → partial patches. | `divide_to_patches`. |
| **`TestResizeAndPadImage`** | Exact target size; landscape/portrait; noop; black padding. | Both scale branches + center paste in `resize_and_pad_image`. |
| **`TestExpand2Square`** | Square noop; pad wide; grayscale → RGB. | All branches in `expand2square`. |
| **`TestUnpadImage` / `TestUnpadImageShape`** | Wide vs tall original → crop; same aspect → full tensor; shape helper matches tensor. | Both aspect branches in `unpad_image` / `unpad_image_shape`. |
| **`TestLoadImageFromBase64`** | Valid PNG round-trip; invalid/empty base64 raises. | `load_image_from_base64` + decode errors. |
| **`TestGetDpEncoderLbAssignment`** | Greedy size balance; empty; single GPU; more GPUs than samples. | Full `get_dp_encoder_lb_assignment`. |
| **`TestProcessAnyresImage`** | Dummy processor: `crop_size` dict path; `crop_size is None` → `shortest_edge` fallback. | Main `process_anyres_image` path + `processor.size` tuple vs dict exception branch. |
| **`TestProcessImages`** | `pad` path stacks; `anyres` stacks when shapes align; **monkeypatched** anyres when per-image shapes differ → **list** branch; default path → `image_processor(images)`. | `process_images`: pad loop, anyres loop + stack vs list, else branch. |

**Uncovered by design (unless added later):** `run_dp_sharded_vision_model`, `run_dp_sharded_mrope_vision_model`.

---

### `test_base_processor.py` → `sglang.srt.multimodal.processors.base_processor` (loaded from disk + import stubs)

| Area | What it checks | Coverage |
|------|----------------|----------|
| **`TestBaseMultiModalProcessorOutput`** | `organize_results` modality order; empty; images-only. | Output dataclass helper only. |
| **`TestMultimodalSpecialTokens`** | Token→modality (exact / regex / no match); `parse_regex`; IDs by modality; combined regex split; cached combined regex. | Special-token helpers only. |
| **`TestValidateMmData`** | All `None` OK; wrong type raises; precomputed must be single item; valid precomputed passes. | `validate_mm_data` / `_validate_one_modality`. |
| **`TestMmOffsets`** | `get_mm_items_offset` runs; `get_mm_items_offset_by_pair` none/single/multiple. | Static offset helpers. |
| **`TestProcessLoadedMmData`** | `_process_loaded_mm_data` for raw image, list, precomputed dict, video, audio. | One private method; not full load/IPC pipeline. |

**Largely uncovered here:** async abstract method, `load_mm_data` / fast & legacy paths, executor + CUDA IPC, hardware-specific branches.

---

### `test_customized_mm_processor_utils.py` → `customized_mm_processor_utils.py`

| Test | What it checks | Coverage |
|------|----------------|----------|
| **Missing `model_type`** | Decorator raises `ValueError`. | Error branch in decorator. |
| **Register by `model_type`** | Registry maps `model_type` → processor class. | Happy path. |
| **Overwrite** | Later registration replaces earlier. | Map overwrite path. |

**Expected:** **~100%** of this file.

---

### `test_internvl_utils.py` → `internvl_utils.py` (skipped if import fails)

| Area | What it checks | Coverage |
|------|----------------|----------|
| **`TestFindClosestAspectRatio`** | Closest ratio; tie-break; empty targets → `(1,1)`; exact match. | `find_closest_aspect_ratio`. |
| **`TestDynamicPreprocess`** | Single block; multi-block + thumbnail; multi-block no thumbnail; tile sizes; very wide image. | `dynamic_preprocess` + thumbnail branch. |
| **`TestImageToPixelValues`** | Stacked tensor shape; thumbnail can increase batch dim; finite `float32`; custom mean/std changes values. | `image_to_pixel_values` + `build_transform`. |

**Expected:** **~100%** when not skipped.

---

### `evs/test_evs_core.py` → `evs/evs_core.py` (loaded by file path)

| Area | What it checks | Coverage |
|------|----------------|----------|
| **`TestComputeRetainedTokensCount`** | `q=0`; high `q` minimum floor; mid-`q` int truncation; single-frame bound. | `compute_retained_tokens_count`. |
| **`TestComputeRetentionMask`** | High `q` keeps first-frame tokens; `q=0` all True; `spatial_merge_size`; 1D bool mask length. | Full mask pipeline (reshape, similarity, sort, top-k). |
| **`TestEvsInputIdHelpers`** | `tokens_per_frame` list sums to retained total; `replace_offsets_with_tokens_per_frame` single-span vs multi-span. | Both offset-replacement branches + `tokens_per_frame`. |

### CODE REVIEW

File: test/registered/unit/multimodal/evs/test_evs_core.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+import importlib.util
+import os
+import pathlib
+import unittest
+
+import torch
+
+from sglang.test.ci.ci_register import register_cpu_ci
+
+register_cpu_ci(est_time=5, suite="stage-a-cpu-only")
```
Comment (ispobock) [inline test/registered/unit/multimodal/evs/test_evs_core.py]: pls change all the registered suite from `stage-a-cpu-only` to `stage-a-test-cpu`.

**Review by ispobock — COMMENTED** [review]
1. Use CustomTestCase instead of unittest.TestCase
2. `importlib.exec_module` is hacky. Three files load modules by file path + exec_module to bypass heavy dependencies. Other UTs in the project use normal imports. Prefer from sglang.srt.multimodal.evs.evs_core import ... with unittest.mock.patch to mock heavy deps.
3. `sys.modules` replacement in test_base_processor.py is risky. It injects ~10 stub modules at collection time (not in setUp/tearDown). If pytest collects other tests in the same process that depend on sglang.srt.utils or transformers, they may see the stubs. Use mock.patch.dict(sys.modules, ...) scoped to the test class instead.
